### PR TITLE
MG-386: Added new input types to rename_columns()

### PR DESCRIPTION
**Pull request by beatrizsaldana**
## Problem

Previously, all transformation results were automatically converted to pandas DataFrames using the `convert_transformation_result_to_dataframe` function before being serialized to JSON. This conversion process created unwanted `None` values when dealing with sparse data structures like those in disease correlation transforms.

For example, in disease correlation data where different models have different modules (IFG, TCX, PHG), converting to a DataFrame would create columns for all possible modules, filling missing values with `None`:

**Before (with None values):**
```json
[
  {
    "model": "Model1",
    "IFG": {"correlation": 0.097, "adj_p_val": 0.389},
    "TCX": {"correlation": 0.100, "adj_p_val": 0.289},
    "PHG": None  // Unwanted None value
  },
  {
    "model": "Model2", 
    "IFG": None,  // Unwanted None value
    "TCX": None,  // Unwanted None value
    "PHG": {"correlation": 0.084, "adj_p_val": 0.409}
  }
]
```

## Solution

The fix removes the automatic DataFrame conversion and preserves the original data structure throughout the pipeline:

1. **Modified `process.py`**: Removed the call to `convert_transformation_result_to_dataframe()` and updated variable names for clarity
2. **Enhanced `load.py`**: Updated function signatures and parameter names to be more descriptive
3. **Improved `utils.py`**: Enhanced the `rename_columns` function to handle lists and dictionaries in addition to DataFrames

#### `src/agoradatatools/process.py`
- Removed automatic DataFrame conversion of transformation results
- Updated variable names from `df` to `transform_result` for clarity
- Preserved original data structure (list, dict, or DataFrame) throughout processing

#### `src/agoradatatools/etl/load.py`
- Updated function signatures to use more descriptive parameter names
- Maintained existing functionality while improving code readability

#### `src/agoradatatools/etl/utils.py`
- Enhanced `rename_columns` function to handle multiple data types (DataFrame, list, dict)
- Added support for renaming keys in dictionaries and lists of dictionaries
- Maintained backward compatibility with existing DataFrame operations

### Result

**After (clean output without None values):**
```json
[
  {
    "model": "Model1",
    "IFG": {"correlation": 0.097, "adj_p_val": 0.389},
    "TCX": {"correlation": 0.100, "adj_p_val": 0.289}
  },
  {
    "model": "Model2",
    "PHG": {"correlation": 0.084, "adj_p_val": 0.409}
  }
]
```

## Testing

- New tests were added to account for new input types for rename_columns function
- All existing tests continue to pass

### CODE REVIEW

File: src/agoradatatools/etl/utils.py
Comment (thomasyu888) [inline src/agoradatatools/etl/utils.py:1]: For guidance, this is the main script to look at, the others are nice linting done by @beatrizsaldana !

File: src/agoradatatools/process.py [hunk 1]
```diff
@@ -153,42 +155,41 @@ def process_dataset(
...
         entities_as_df[entity_name] = df
 
     if "custom_transformations" in dataset_obj[dataset_name].keys():
-        result = apply_custom_transformations(
+        transform_result = apply_custom_transformations(
             datasets=entities_as_df,
             dataset_name=dataset_name,
             dataset_obj=dataset_obj[dataset_name],
         )
-        df = utils.convert_transformation_result_to_dataframe(result, dataset_name)
```
Comment (thomasyu888) [inline src/agoradatatools/process.py:167]: For review guidance: this changed as well.

File: src/agoradatatools/etl/utils.py [hunk 2]
```diff
@@ -106,23 +105,40 @@ def standardize_values(df: pd.DataFrame) -> pd.DataFrame:
...
+        data (pd.DataFrame, list, dict): DataFrame, list of dictionaries, or dictionary with columns to be renamed
         column_map (dict): Dictionary mapping original column names to new columns
 
     Returns:
-        pd.DataFrame: DataFrame with new columns names
+        pd.DataFrame, list, dict: DataFrame, list of dictionaries, or dictionary with new columns names
     """
+
+    def _rename_dict(d: dict, column_map: dict) -> None:
+        """
```
Comment (linglp) [inline src/agoradatatools/etl/utils.py:123]: Nit: can we make the docstring more comprehensive here?

File: src/agoradatatools/etl/utils.py [hunk 3]
```diff
@@ -106,23 +105,40 @@ def standardize_values(df: pd.DataFrame) -> pd.DataFrame:
...
+                d[new_key] = d.pop(old_key)
+
     try:
-        df.rename(columns=column_map, inplace=True)
+        if isinstance(data, list):
+            for item in data:
+                _rename_dict(item, column_map)
+        elif isinstance(data, dict):
+            _rename_dict(data, column_map)
+        else:
```
Comment (linglp) [inline src/agoradatatools/etl/utils.py]: Nit: I think if data is not one of: list, dict or dataframe, there should be an error raised like below? 
```
# Unsupported type
    raise TypeError(
        f"Unsupported data type {type(data).__name__}; expected dict, list[dict], or pandas.DataFrame."
    )
```

File: src/agoradatatools/etl/utils.py [hunk 4]
```diff
@@ -106,23 +105,40 @@ def standardize_values(df: pd.DataFrame) -> pd.DataFrame:
...
+            data.rename(columns=column_map, inplace=True)
     except TypeError:
```
Comment (linglp) [inline src/agoradatatools/etl/utils.py]: Nit: Checking the type of column mapping can probably be moved to earlier? 
```
from collections.abc import Mapping
    if not isinstance(column_map, Mapping):
        raise TypeError("column_map must be a mapping (e.g., dict) of {old_key: new_key}.")
```

File: src/agoradatatools/process.py [hunk 5]
```diff
@@ -153,42 +155,41 @@ def process_dataset(
...
-    if isinstance(df, dict):
+    if isinstance(transform_result, dict):
         json_path = load.dict_to_json(
-            df=df,
+            data_as_dict=transform_result,
             staging_path=staging_path,
             filename=dataset_name + "." + dataset_obj[dataset_name]["final_format"],
         )
-    elif isinstance(df, list):
+    elif isinstance(transform_result, list):
```
Comment (linglp) [inline src/agoradatatools/process.py:184]: Nit: do we have test for this part? I saw `test_dict_to_json_success` but not anything related to `test_list_to_json`

File: tests/test_utils.py [hunk 6]
```diff
@@ -190,19 +190,152 @@ class TestRenameColumns:
...
         captured_output = StringIO()
         sys.stdout = captured_output
         bad_renamed_df = utils.rename_columns(
-            df=self.df.copy(), column_map=self.bad_column_map
+            data=self.df.copy(), column_map=self.bad_column_map
         )
         assert "Column mapping must be a dictionary" in captured_output.getvalue()
         assert list(bad_renamed_df.columns) == list(self.good_column_map.keys())
 
+    def test_rename_columns_with_dict_input(self):
```
Comment (linglp) [inline tests/test_utils.py]: Nit: I think all these tests are great, but you probably want to group them based on input types? Here's an example: 
```
class TestRenameColumnsDict:
    def test_basic(self): ...
    def test_partial_mapping(self): ...
    def test_nonexistent_keys(self): ...
    def test_preserves_values(self): ...
    def test_complex_nested(self): ...

class TestRenameColumnsList:
    def test_basic(self): ...
    def test_empty_list(self): ...
    def test_mixed_list_content(self): ...

class TestRenameColumnsDataFrame:
    def test_preserves_dataframe_values(self): ...
```

**Review by linglp — APPROVED** [review]
Thanks for your work, @beatrizsaldana ! The changes here make sense to me. If I understand correctly, previously we always converted the results of a custom transformation into a DataFrame and then renamed the columns. With your changes, the renaming now happens directly based on the input type returned by the custom transformation function.

File: src/agoradatatools/etl/utils.py (hunk 3 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: I wonder if it might get raised automatically because of the type checking in the function arguments? I'm a little fuzzy on that part of Python.

File: src/agoradatatools/etl/utils.py [hunk 7]
```diff
@@ -106,23 +105,40 @@ def standardize_values(df: pd.DataFrame) -> pd.DataFrame:
...
+        """
+        Rename keys in a dictionary in place.
+        """
+        for old_key, new_key in column_map.items():
+            if old_key in d:
+                d[new_key] = d.pop(old_key)
+
     try:
-        df.rename(columns=column_map, inplace=True)
+        if isinstance(data, list):
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/utils.py]: One potential issue: This only checks if data is a `list`, _not_ that each item in the list is a `dict`. Possibly this check could be done inside the `for old_key, new_key...` loop in `_rename_dict` so each item gets checked. 

OR... does enforcing `d: dict` in the function arguments auto-check that?

File: src/agoradatatools/process.py (hunk 5 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/process.py:184]: +1, I don't see one in the file

File: tests/test_utils.py (hunk 6 above)
Comment (jaclynbeck-sage) [reply tests/test_utils.py]: I like this idea!

File: tests/test_utils.py [hunk 8]
```diff
@@ -190,19 +190,152 @@ class TestRenameColumns:
...
+        """Test renaming columns in a dictionary input"""
+        input_dict = {"a": "value1", "b": "value2", "c": "value3"}
+        column_map = {"a": "x", "b": "y", "c": "z"}
+
+        result = utils.rename_columns(data=input_dict, column_map=column_map)
+
+        expected = {"x": "value1", "y": "value2", "z": "value3"}
+        assert result == expected
+        # Verify the original dict was modified in place
+        assert input_dict == expected
```
Comment (jaclynbeck-sage) [inline tests/test_utils.py]: Interesting... if things are modified in place, can you add that to the docstring for `rename_columns` function? It's not clear that that's the behavior for dicts from the code.

File: tests/test_utils.py [hunk 9]
```diff
@@ -190,19 +190,152 @@ class TestRenameColumns:
...
+        column_map = {"a": "x", "b": "y", "c": "z"}
+
+        result = utils.rename_columns(data=input_dict, column_map=column_map)
+
+        expected = {
+            "x": {"nested": {"deep": "value"}},
+            "y": [{"inner": "data"}, {"inner": "more_data"}],
+            "z": "simple_string",
+        }
+        assert result == expected
```
Comment (jaclynbeck-sage) [inline tests/test_utils.py]: This is a great set of tests! Can you also add tests for some expected _error_ conditions:
1. Column map is not a dict
2. Column map is a dict but contains values that aren't valid renames like {"a": []} and {"a": None}
3. Input is not a data.frame, list, or dict
4. Input is a list but not a list of dicts
5. Input is a list of mixed types: some are dicts and some are something else

File: tests/test_utils.py [hunk 10]
```diff
@@ -190,19 +190,152 @@ class TestRenameColumns:
...
+        """Test renaming with an empty column mapping"""
+        input_dict = {"a": "value1", "b": "value2"}
+        column_map = {}
+
+        result = utils.rename_columns(data=input_dict, column_map=column_map)
+
+        # Should return unchanged data
+        assert result == {"a": "value1", "b": "value2"}
+
+    def test_rename_columns_with_empty_list(self):
```
Comment (jaclynbeck-sage) [inline tests/test_utils.py:367]: Can you add basically this same test but for input = empty dict {} and empty DataFrame()?

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
This is really great work. I just need a few extra comments added and some tests that intentionally fail.

File: src/agoradatatools/etl/utils.py (hunk 7 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/utils.py]: Python type hints are purely informational, they don't prevent the function from running with incorrect types. They're like documentation that tools can use, but Python itself ignores them during execution. We would have to configure a repo-wide runtime type-checking tool in order to prevent the code from running using the type hints. So, we need the `if isinstance()` stuff.

Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: Got it! Looks like you changed it to `list[dict]`, so we're good here.

**Review by jaclynbeck-sage — COMMENTED** [review]
I think everything looks good except, is there a test for this scenario?
Column map is a dict but contains values that aren't valid renames like {"a": []} and {"a": None}

**Review by jaclynbeck-sage — APPROVED** [review]
Looks great!
